ni_measurementlink_service/session_management/_reservation.py

```diff
@@ -94,6 +142,52 @@ def _grpc_channel_pool(self) -> GrpcChannelPool:
             raise ValueError("This method requires a gRPC channel pool.")
         return self._session_manager._grpc_channel_pool
 
+    @cached_property
+    def _reserved_pin_or_relay_names(self) -> Iterable[str]:
+        # Use dict keys to preserve insertion order.
+        names = {}
+        for session_info in self._session_info:
+            for channel_mapping in session_info.channel_mappings:
+                names[channel_mapping.pin_or_relay_name] = True
+        return list(names.keys())
+
+    @cached_property
+    def _reserved_sites(self) -> Iterable[int]:
```

Comment:
In LabVIEW, the pin map context has an array of sites and I use that as the order. Similarly, for the pin order, I just cache off the array of pins passed when reserving the session. I don't have enough intuition on the order of the channel mappings, but I suspect you won't necessarily get the same ordering with this logic as we do in LabVIEW. It also seems like it would be difficult for a user to intuit the order based on this logic.

Did you do it this way because it allows you to also order by instrument type id in a consistent way? If so, I might prefer that instrument type id ordering should either be non guaranteed or simply be alphabetical order. We don't run into this problem in LabVIEW (yet) because we always pass an instrument type id. When we implement the generic Get Connections, I haven't thought about whether the instrument type id should be required or not. If not, then we'll have to tackle ordering in LabVIEW too.

Reply:
After looking at the tests, changing this will have some pretty serious ramifications on the tests. But I suspect refactoring them in python will be a lot easier than in LabVIEW.  :D

Reply:
This is the fallback order for reserve_all_registered_sessions.

Reply:
I must be missing something. The way I read the code below, if I don't pass sites to get_connections(), the site order will get populated with the result of _reserved_sites(). Same with pins.

Reply:
_reserved_sites is used if you don't specify sites when calling get_connection(s). However, it doesn't always come from this `@cached_property` function.
- reserve_sessions initializes _reserved_sites based on the pin map context, so it doesn't use the cached property.
- reserve_all_registered_sessions does not initialize _reserved_sites, so it uses the cached property.

Reply:
Thanks. I got it now.

Reply:
I think this was just my confusion about how `@cached_property` works. Upon looking at the docs, what you are doing now makes sense.

Reply:
I admit that it's subtle and not obvious if you're not familiar with `@cached_property`. That's why I wrote the comments about "If __init__ doesn't initialize ..., the cached properties lazily initialize them" and "If ...  is not specified at construction time ...".